src/Client/ClientBase.cpp

```diff
@@ -1892,48 +1893,19 @@ void ClientBase::processParsedSingleQuery(const String & full_query, const Strin
     profile_events.watch.restart();
 
     {
-        /// Temporarily apply query settings to context.
-        std::optional<Settings> old_settings;
-        SCOPE_EXIT_SAFE({
-            if (old_settings)
-                client_context->setSettings(*old_settings);
-        });
-
-        auto apply_query_settings = [&](const IAST & settings_ast)
+        /// Temporarily apply query settings to the context.
+        Settings old_settings = client_context->getSettingsCopy();
+        SCOPE_EXIT_SAFE(
         {
-            if (!old_settings)
-                old_settings.emplace(client_context->getSettingsRef());
-            client_context->applySettingsChanges(settings_ast.as<ASTSetQuery>()->changes);
-            client_context->resetSettingsToDefaultValue(settings_ast.as<ASTSetQuery>()->default_settings);
-        };
-
-        const auto * insert = parsed_query->as<ASTInsertQuery>();
-        if (const auto * select = parsed_query->as<ASTSelectQuery>(); select && select->settings())
-            apply_query_settings(*select->settings());
-        else if (const auto * select_with_union = parsed_query->as<ASTSelectWithUnionQuery>())
-        {
-            const ASTs & children = select_with_union->list_of_selects->children;
-            if (!children.empty())
-            {
-                // On the client it is enough to apply settings only for the
-                // last SELECT, since the only thing that is important to apply
-                // on the client is format settings.
-                const auto * last_select = children.back()->as<ASTSelectQuery>();
-                if (last_select && last_select->settings())
-                {
-                    apply_query_settings(*last_select->settings());
-                }
-            }
-        }
-        else if (const auto * query_with_output = parsed_query->as<ASTQueryWithOutput>(); query_with_output && query_with_output->settings_ast)
-            apply_query_settings(*query_with_output->settings_ast);
-        else if (insert && insert->settings_ast)
-            apply_query_settings(*insert->settings_ast);
+            client_context->setSettings(old_settings);
```

Comment:
After some debugging, I suspect `00653_verification_monotonic_data_load` timeouts under tsan caused by this line, client hung here for some reason. However, no explanation why it reproduced only at one test and didn't reproduce before, even though we had `setSettings` under `if (old_settings)`

Reply:
Also, the issue might be related to grep -q, such as early termination of the client leading to that

Reply:
There’s a data race that this PR reveals
https://github.com/ClickHouse/ClickHouse/issues/68626#issuecomment-2305120194

We can merge this PR after fixing it by https://github.com/ClickHouse/ClickHouse/pull/68791

Reply:
https://github.com/ClickHouse/ClickHouse/pull/69186 is merged, are we missing anything else?